File: gcj_2016/QR_A.py
#!/usr/bin/env python

# Google Code Jam 2016 Qualification Round
# Problem A. Counting Sheep

import sys

def gen(N):
  n = 0
  while True:
    n += N
    yield n

def f(N, d):
  g = gen(N)
  while True:
    n = next(g)
    for a in list(str(n)):
      if int(a) == d:
        return n

def solve(N):
  if N == 0:
    return "INSOMNIA"
  return max([f(N, d) for d in range(0, 10)])

T = int(sys.stdin.readline())
for t in range(1, T+1):
  N = int(sys.stdin.readline())
  ans = solve(N)
  print("Case #" + str(t) + ": " + str(ans))

File: gcj_2016/QR_A.pl
#!/usr/bin/env perl

# Google Code Jam 2016 Qualification Round
# Problem A. Counting Sheep

use List::Util 'max';

sub solve {
  my $n = $_[0];
  return "INSOMNIA" if $n == 0;

  max map {
    my $d = ''.$_, $x;
    for ($x = $n; "$x" !~ /$d/; $x += $n) {
    }
    $x;
  } (0..9);
}

my $T = <>;
for (my $t = 1; $t <= $T; $t++) {
  my $n = <> + 0;
  my $ans = solve($n);
  print "Case #$t: $ans\n";
}
